### PR TITLE
Add an initial benchmark for the tx-parser crate. (#406) r=nalexander

### PR DESCRIPTION
**Pull request by bgrins**
This is an example first benchmark for the tx-parser.  It can be run with:

```cargo bench --package mentat_tx_parser```

To take measurements and print it in a nicer format I used this command:

```for i in `seq 1 5`; do echo '---'; cargo bench --package mentat_tx_parser 2> /dev/null | sed -n -e 's/test \(.*\)\( ... bench: \)/\1/p'; done;```

Here are my results for that locally:

    bench_parse1      7,462 ns/iter (+/- 7,514)
    bench_parse2    384,068 ns/iter (+/- 48,037)
    ---
    bench_parse1      7,171 ns/iter (+/- 1,431)
    bench_parse2    381,971 ns/iter (+/- 59,735)
    ---
    bench_parse1      6,860 ns/iter (+/- 1,742)
    bench_parse2    352,454 ns/iter (+/- 177,286)
    ---
    bench_parse1      6,787 ns/iter (+/- 3,107)
    bench_parse2    348,041 ns/iter (+/- 146,625)
    ---
    bench_parse1      7,113 ns/iter (+/- 961)
    bench_parse2    376,373 ns/iter (+/- 52,412)

### CODE REVIEW

**Comment by rnewman** [comment]
lol, I think that script doesn't like zsh:

```
$ for i in `seq 1 5`; do echo '---'; cargo bench --package mentat_tx_parser 2> /dev/null | sed -n -e 's/test \(.*\)\( ... bench: \)/\1/p'; done;
---
---
---
---
---
```

**Comment by rnewman** [comment]
Oh, it fails 'cos I'm not using Nightly.

**Comment by rnewman** [comment]
Older dual-core 13" rMBP:

```
test bench_parse1 ... bench:       8,959 ns/iter (+/- 1,450)
test bench_parse2 ... bench:     594,676 ns/iter (+/- 66,751)
---
test bench_parse1 ... bench:       8,634 ns/iter (+/- 4,373)
test bench_parse2 ... bench:     575,320 ns/iter (+/- 71,528)
```

File: tx-parser/benches/bench.rs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+use test::Bencher;
+use mentat_tx_parser::Tx;
+
+#[bench]
+fn bench_parse1(b: &mut Bencher) {
+    let input = r#"[[:db/add 1 :test/val "a"]]"#;
+    let parsed_edn = edn::parse::value(input).expect("to parse test input");
+    b.iter(|| {
+        return Tx::parse(parsed_edn.clone());
```
Comment (rnewman) [inline tx-parser/benches/bench.rs]: Nit: drop `return` and `;` (and below).

Concern: we're cloning the string here, inside the benchmark. Is it fairer to clone it outside? Or is it more realistic to call `edn::parse` inside the loop, and not explicitly clone?

Comment (bgrins) [inline tx-parser/benches/bench.rs]: > Concern: we're cloning the string here, inside the benchmark. Is it fairer to clone it outside?

That seems ideal, but without the clone we get this error:
```
53 |         return Tx::parse(parsed_edn);
   |                          ^^^^^^^^^^ cannot move out of captured outer variable in an `FnMut` closure
```

> Or is it more realistic to call edn::parse inside the loop, and not explicitly clone?

We could - I thought cloning the parsed edn would be more narrow than doing the actual edn parsing in the loop.  If we wanted to fold edn parsing into this benchmark that'd be OK but it turns into a broader test - I figured we might want a separate benchmark for just parsing out the EDN.

With my test script and edn parsing in the loop I get:

```
bench_parse1      6,814 ns/iter (+/- 2,250)
bench_parse2    604,382 ns/iter (+/- 70,221)
---
bench_parse1      7,526 ns/iter (+/- 2,517)
bench_parse2    631,710 ns/iter (+/- 86,643)
---
bench_parse1      7,073 ns/iter (+/- 2,645)
bench_parse2    624,657 ns/iter (+/- 74,293)
---
bench_parse1      7,124 ns/iter (+/- 2,509)
bench_parse2    636,105 ns/iter (+/- 83,272)
---
bench_parse1      7,118 ns/iter (+/- 3,830)
bench_parse2    620,871 ns/iter (+/- 116,928)
```

And with the clone in the loop and parsing outside I get:

```
---
bench_parse1      6,925 ns/iter (+/- 1,087)
bench_parse2    382,473 ns/iter (+/- 63,815)
---
bench_parse1      7,073 ns/iter (+/- 1,339)
bench_parse2    428,671 ns/iter (+/- 185,621)
---
bench_parse1      7,392 ns/iter (+/- 2,354)
bench_parse2    398,395 ns/iter (+/- 134,769)
---
bench_parse1      7,044 ns/iter (+/- 792)
bench_parse2    394,561 ns/iter (+/- 44,045)
---
bench_parse1      6,685 ns/iter (+/- 1,071)
bench_parse2    387,890 ns/iter (+/- 70,388)
```

Comment (bgrins) [inline tx-parser/benches/bench.rs]: Note: on those numbers I had only made the change to bench_parse2 so ignore bench_parse1

File: tx-parser/benches/bench.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+         [:db/add 20 :test/val "t"]
+         [:db/add 21 :test/val "u"]
+         [:db/add 22 :test/val "v"]
+         [:db/add 23 :test/val "w"]
+         [:db/add 24 :test/val "x"]
+         [:db/add 25 :test/val "y"]
+         [:db/add 26 :test/val "z"]]"#;
+    let parsed_edn = edn::parse::value(input).expect("to parse test input");
+    b.iter(|| {
+        return Tx::parse(parsed_edn.clone());
```
Comment (ncalexan) [inline tx-parser/benches/bench.rs]: As @rnewman suggested, just `|| Tx::parse(...)` is idiomatic.

**Pull request merged**